src/qt/listctrl.cpp

```diff
@@ -141,96 +144,868 @@ class wxQtTreeWidget : public wxQtEventSignalHandler< QTreeWidget, wxListCtrl >
-        event.m_item.SetId(event.m_itemIndex);
-        event.m_item.SetMask(wxLIST_MASK_TEXT |
-                             wxLIST_MASK_IMAGE |
-                             wxLIST_MASK_DATA);
-        handler->GetItem(event.m_item);
-        EmitEvent(event);
+        return m_rowCount;
     }
-}
 
-void wxQtTreeWidget::itemClicked(QTreeWidgetItem *qitem, int column)
-{
-    EmitListEvent(wxEVT_LIST_ITEM_SELECTED, qitem, column);
-}
+    QVariant data(const QModelIndex &index, int role) const wxOVERRIDE
+    {
+        wxListCtrl *listCtrl = GetListCtrl();
 
-void wxQtTreeWidget::itemPressed(QTreeWidgetItem *qitem, int column)
-{
-    EmitListEvent(wxEVT_LIST_ITEM_SELECTED, qitem, column);
-}
+        const int row = index.row();
+        const int col = index.column();
 
-void wxQtTreeWidget::itemActivated(QTreeWidgetItem *qitem, int column)
-{
-    EmitListEvent(wxEVT_LIST_ITEM_ACTIVATED, qitem, column);
-}
+        if ( role == Qt::DisplayRole || role == Qt::EditRole )
+        {
+            const wxString text = listCtrl->OnGetItemText(row, col);
+            return QVariant::fromValue(wxQtConvertString(text));
+        }
 
+        if ( role == Qt::DecorationRole )
+        {
+            wxImageList *imageList = GetImageList();
+            if ( imageList == NULL )
+                return QVariant();
+
+            const int imageIndex = listCtrl->OnGetItemColumnImage(row, col);
+            if (imageIndex == -1)
+                return QVariant();
+            wxBitmap image = imageList->GetBitmap(imageIndex);
+            wxCHECK_MSG(image.IsOk(), QVariant(), "Invalid Bitmap");
+            return QVariant::fromValue(*image.GetHandle());
+        }
 
-Qt::AlignmentFlag wxQtConvertTextAlign(wxListColumnFormat align)
-{
-    switch (align)
+        return QVariant();
+    }
+
+    virtual bool GetItem(wxListItem& info) wxOVERRIDE
     {
-        case wxLIST_FORMAT_LEFT:
-            return Qt::AlignLeft;
-        case wxLIST_FORMAT_RIGHT:
-            return Qt::AlignRight;
-        case wxLIST_FORMAT_CENTRE:
-            return Qt::AlignCenter;
+        const int row = static_cast<int>(info.GetId());
+        const int col = info.m_col;
+
+        if (info.m_mask & wxLIST_MASK_TEXT)
+            info.SetText(GetListCtrl()->OnGetItemText(row,col));
+
+        CopySelectStatusToItem(info, row, col);
+
+        return true;
     }
-    return Qt::AlignLeft;
-}
 
-wxListColumnFormat wxQtConvertAlignFlag(int align)
-{
-    switch (align)
+    bool IsVirtual() const wxOVERRIDE
     {
-        case Qt::AlignLeft:
-            return wxLIST_FORMAT_LEFT;
-        case Qt::AlignRight:
-            return wxLIST_FORMAT_RIGHT;
-        case Qt::AlignCenter:
-            return wxLIST_FORMAT_CENTRE;
+        return true;
     }
-    return wxLIST_FORMAT_LEFT;
-}
 
+    void SetVirtualItemCount(long count) wxOVERRIDE
+    {
+        const int last = count > 0 ? count - 1 : 0;
+        beginInsertRows(QModelIndex(), 0, last);
```

Comment:
Do we really need to do this even though we're not actually inserting anything here?

Reply:
The view is only updated when the model changes. So we do need to tell QT that something has changed.  Conceptually, we are changing the number of rows in the model here.